### PR TITLE
Add optional sort and dir params to host API

### PR DESCRIPTION
**Pull request by kahowell**
To test, use swagger-UI with some hosts inserted via
`bin/insert-mock-hosts`.

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/resource/HostsResource.java [hunk 1]
```diff
@@ -53,16 +59,36 @@
...
+        Sort sortValue = Sort.unsorted();
+        Sort.Direction dirValue = Sort.Direction.ASC;
+
+        if (dir == SortDirection.DESC) {
+            dirValue = Sort.Direction.DESC;
+        }
+        if (sort != null) {
+            sortValue = Sort.by(dirValue, sortParamMapping.get(sort));
+        }
+
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/resource/HostsResource.java:92]: Had to add an exception mapper for BadRequestException.

**Comment by kahowell** [comment]
I realized this needs to incorporate changes in #157, rebased accordingly, and I amended these changes in:

```diff
diff --git a/src/test/java/org/candlepin/subscriptions/resource/HostsResourceTest.java b/src/test/java/org/candlepin/subscriptions/resource/HostsResourceTest.java
index 3c02489..b58badb 100644
--- a/src/test/java/org/candlepin/subscriptions/resource/HostsResourceTest.java
+++ b/src/test/java/org/candlepin/subscriptions/resource/HostsResourceTest.java
@@ -24,8 +24,12 @@ import static org.mockito.Mockito.*;
 
 import org.candlepin.subscriptions.db.HostRepository;
 import org.candlepin.subscriptions.db.model.Host;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.Usage;
 import org.candlepin.subscriptions.resteasy.PageLinkCreator;
 import org.candlepin.subscriptions.security.WithMockRedHatPrincipal;
+import org.candlepin.subscriptions.tally.AccountListSource;
+import org.candlepin.subscriptions.tally.AccountListSourceException;
 import org.candlepin.subscriptions.utilization.api.model.HostReportSort;
 import org.candlepin.subscriptions.utilization.api.model.SortDirection;
 
@@ -52,14 +56,18 @@ class HostsResourceTest {
     @MockBean
     PageLinkCreator pageLinkCreator;
 
+    @MockBean
+    AccountListSource accountListSource;
+
     @Autowired
     HostsResource resource;
 
     @BeforeEach
-    public void setup() {
+    public void setup() throws AccountListSourceException {
         PageImpl<Host> mockPage = new PageImpl<>(Collections.emptyList());
         when(repository.getHostsByBucketCriteria(any(), any(), any(), any(), any(), any(), any()))
             .thenReturn(mockPage);
+        when(accountListSource.containsReportingAccount("account123456")).thenReturn(true);
     }
 
     @Test
@@ -69,8 +77,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "displayName"))
@@ -84,8 +92,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "cores"))
@@ -99,8 +107,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "sockets"))
@@ -114,8 +122,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "lastSeen"))
@@ -129,8 +137,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "hardwareMeasurementType"))
@@ -144,8 +152,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1))
@@ -159,8 +167,8 @@ class HostsResourceTest {
         verify(repository, only()).getHostsByBucketCriteria(
             eq("account123456"),
             eq("RHEL"),
-            eq(null),
-            eq(null),
+            eq(ServiceLevel.ANY),
+            eq(Usage.ANY),
             eq(null),
             eq(false),
             eq(PageRequest.of(0, 1, Sort.Direction.ASC, "displayName"))
```

I'll mark as draft until #157 lands.

**Comment by kahowell** [comment]
5505b6a to 4e8bd10: rebase on master.

Depends on #157... that commit cherry-picked here for convenience, please keep as WIP until that is merged, at that time, I will rebase it out.

**Comment by mstead** [comment]
@kahowell #157 now merged. Should now be able to move out of draft.

**Pull request merged**